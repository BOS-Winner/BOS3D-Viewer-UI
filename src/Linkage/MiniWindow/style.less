.mini-window {
  position: relative;
  width: 100%;
  height: 100%;
  overflow: hidden;
  border-radius: 4px;
  background: rgba(27, 27, 27, 0.9);
  border-radius: 4px;
}

.title {
  box-sizing: border-box;
  display: flex;
  height: 48px;
  justify-content: space-between;
  align-items: center;
  width: calc(100%);
  padding: 0 10px;
  color: white;
  background: rgba(27, 27, 27, 0.95);
  border-radius: 4px 4px 0 0;
  border: 1px solid #575758;
  .name{
    display: flex;
    align-items: center;
    >img{
      margin-right: 7px;
    }
  }
  .center-title{
    display: inline-block;
    position: absolute;
    transform: translate(-50%, 0);
    left: 50%;
  }
}


.container {
  position: relative;
  width: 100%;

  .toolbar {
    position: absolute;
    bottom: 10px;
    left: 50%;
    transform: translate(-50%, 0);
  }
  .mobile-toolbar {
    position: absolute;
    bottom: 16px;
    left: 50%;
    transform: translate(-50%, 0);
  }
}

.icon-close {
  cursor: pointer;
  @size: 16px;
  SVG{
    width: @size;
    height: @size;
  }
}
