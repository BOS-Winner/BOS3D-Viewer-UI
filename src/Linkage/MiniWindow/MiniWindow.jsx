import React, {
  useRef, useImperativeHandle, forwardRef
} from 'react';
import PropTypes from "prop-types";
import Toolbar from "Base/Toolbar";
import Icon from "Base/Icon";
import titleimg from "Base/Modal/img/titleIcon.png";
import CONSTANTS from "../constant";
import icon3D from "../img/3D.png";
import icon2D from "../img/2D.png";
import style from "./style.less";
import { AntdIcon, mobileCheck } from '../../UI/utils/utils';
import iconStyle from '../../BOS2DUI/Theme/icon.less';

const TITLE_HEIGHT = CONSTANTS.TITLE_HEIGHT;

/**
 * 显示小窗口
 * @param {object} props
 * @param {string} [props.className=''] - className
 * @param {CSSStyleDeclaration} [props.style={}] - style
 * @param {string} [props.title=''] - title
 * @param {HTMLElement} props.children - 子元素
 * @param {'' | '3D' | '2D'} [props.icon=''] - 底部需要什么图标
 * @param {boolean} props.showTitle - 是否显示标题栏
 * @param {boolean} [props.showCloseIcon = true] - 是否显示关闭按钮
 * @param {function} [props.onClose = () => {}] - 点击关闭按钮的回调
 * @return {object}
 * @constructor
 */
function MiniWindow(props, ref) {
  let img = icon2D;
  if (props.icon === '3D') {
    img = icon3D;
  }
  const isMobile = mobileCheck();
  const dom = useRef(null);

  useImperativeHandle(ref, () => ({
    reset: () => {
      dom.current.style.zIndex = 1;
      if (isMobile) document.querySelector('#mobileToolbar').style.display = 'block';
    }
  }));

  // PC端工具条
  const PCTool = (
    <Toolbar className={style.toolbar}>
      {props.icon === '2D' && (
        <div title="框选">
          <Icon
            className={iconStyle.icon}
            icon={<AntdIcon type="iconboxselection" />}
            title="框选"
            onClick={() => props.onClickPick()}
          />
        </div>
      )}
      {props.icon === '2D' && (
        <div title="图层">
          <Icon
            className={iconStyle.icon}
            icon={<AntdIcon type="icontuceng2-01" />}
            title="图层"
            onClick={() => props.onLayer()}
          />
        </div>
      )}
      {props.icon && (
        <div title={props.icon === '3D' ? '切换3D视图' : "切换2D视图"}>
          <Icon
            className={iconStyle.icon}

            icon={<AntdIcon type="icona-2dshituquanping-01" />}
            title={props.icon === "3D" ? '2D' : '3D'}
            onClick={() => props.onClickMainIcon()}
          />
        </div>
      )}
      {props.icon === '2D' && (
        <div title="分屏">
          <Icon
            className={iconStyle.icon}
            icon={<AntdIcon type="iconfenping2-01" />}
            title="分屏"
            onClick={() => props.onClickSideIcon()}
          />
        </div>
      )}
    </Toolbar>
  );
  const MobileTool = (
    <div>
      <Toolbar className={style.mobileToolbar}>
        {/* {props.icon === '2D' && (
          <div title="图层">
            <Icon
              className={iconStyle.icon}
              icon={<AntdIcon type="icontuceng2-01" />}
              onClick={() => props.onLayer()}
            />
          </div>
        )} */}
        {props.icon && (
          <div title={props.icon === '3D' ? '切换3D视图' : "切换2D视图"}>
            <Icon
              className={iconStyle.icon}
              icon={<AntdIcon type="icona-2dshituquanping-01" />}
              onClick={() => {
                props.onClickMainIcon();
                if (props.icon === '3D') document.querySelector('#mobileToolbar').style.display = 'block';
              }}
            />
          </div>
        )}
        {props.icon === '2D' && (
          <div title="分屏">
            <Icon
              className={iconStyle.icon}
              icon={<AntdIcon type="iconfenping2-01" />}
              onClick={() => props.onClickSideIcon()}
            />
          </div>
        )}
      </Toolbar>
    </div>
  );
  return (
    <div
      className={`${style.miniWindow} ${props.className}`}
      style={props.style}
      ref={dom}
    >
      {props.showTitle && (
        <div
          className={style.title}
        >
          {
            isMobile && props.showCloseIcon && props.icon !== '3D' && props.title === "二维图纸" && props.icon !== "2D" ? (
              <span className={style.centerTitle}>{props.title}</span>
            ) : (
              <div className={style.name}>
                <img src={titleimg} alt="图片" />
                {props.title}
              </div>
            )
          }

          {isMobile && props.showCloseIcon && props.icon !== '3D' && props.title === "二维图纸" && props.icon !== "2D" ? (
            <AntdIcon
              type="iconback"
              className={style.iconClose}
              style={{
                position: 'absolute',
                left: '12px'
              }}
              onClick={ev => {
                ev.stopPropagation();
                props.onClose();
              }}
            />
          ) : (
            isMobile && props.icon === "2D" && props.title === "二维图纸" && (
              <AntdIcon
                type="iconclose"
                className={style.iconClose}
                onClick={ev => {
                  ev.stopPropagation();
                  props.onClose();
                }}
              />
            )
          )}
          {!isMobile && props.title === "二维图纸" && (
            <AntdIcon
              type="iconclose"
              className={style.iconClose}
              onClick={ev => {
                ev.stopPropagation();
                props.onClose();
              }}
            />
          )}
          {props.icon === '3D' && isMobile && (
            <AntdIcon
              type="iconclose"
              className={style.iconClose}
              onClick={ev => {
                ev.stopPropagation();
                // setShow(false);
                dom.current.style.zIndex = -1;
              }}
            />
          )}
        </div>
      )}
      <div
        className={style.container}
        style={{
          height: props.showTitle ? `calc(100% - ${TITLE_HEIGHT})` : '100%',
        }}
      >
        {props.children}

        {/* 工具条 */}
        {
          !isMobile ? PCTool : MobileTool
        }

      </div>
    </div>
  );
}

MiniWindow.propTypes = {
  className: PropTypes.string,
  style: PropTypes.object,
  title: PropTypes.string,
  children: PropTypes.node.isRequired,
  icon: PropTypes.oneOf(['', '2D', '3D']),
  showTitle: PropTypes.bool.isRequired,
  showCloseIcon: PropTypes.bool,
  onClose: PropTypes.func,
  onClickMainIcon: PropTypes.func,
  onClickSideIcon: PropTypes.func,
  onClickPick: PropTypes.func,
  onLayer: PropTypes.func,
};

MiniWindow.defaultProps = {
  className: '',
  title: '',
  style: {},
  icon: '',
  showCloseIcon: true,
  onClose: () => { },
  onClickMainIcon: () => { },
  onClickSideIcon: () => { },
  onClickPick: () => { },
  onLayer: () => {},
};

export default forwardRef(MiniWindow);
