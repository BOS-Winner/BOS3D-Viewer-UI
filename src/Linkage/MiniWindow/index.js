import MiniWindow from "./MiniWindow";

export default MiniWindow;
