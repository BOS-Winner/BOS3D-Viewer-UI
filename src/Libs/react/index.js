import React from "react";
import useUpdateEffect from "./useUpdateEffect";

React.useUpdateEffect = useUpdateEffect;

export default React;
// export { default as useUpdateEffect } from "./useUpdateEffect";
