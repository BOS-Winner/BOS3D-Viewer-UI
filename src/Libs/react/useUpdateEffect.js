import { useEffect, useRef } from "react";

export default function (effect, deps) {
  const isFirstRender = useRef(true);
  useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
    } else {
      effect();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}
