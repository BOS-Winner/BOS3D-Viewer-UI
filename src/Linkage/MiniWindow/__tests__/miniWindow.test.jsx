import React from 'react';
import { render } from "@testing-library/react";
import MiniWindow from '../MiniWindow';

describe('miniWindow', () => {
  it('should render dom correctly', () => {
    const { container, rerender } = render((
      <MiniWindow icon="3D" showTitle>
        <div>test</div>
      </MiniWindow>
    ));
    expect(container).toMatchSnapshot();

    rerender((
      <MiniWindow icon="3D" showTitle showCloseIcon={false}>
        <div>test</div>
      </MiniWindow>
    ));
    expect(container).toMatchSnapshot();

    rerender((
      <MiniWindow title="test" icon="3D" showTitle showCloseIcon={false}>
        <div>test</div>
      </MiniWindow>
    ));
    expect(container).toMatchSnapshot();
  });
});
