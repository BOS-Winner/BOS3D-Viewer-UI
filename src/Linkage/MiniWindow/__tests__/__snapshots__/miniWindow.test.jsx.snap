// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`miniWindow should render dom correctly 1`] = `
<div>
  <div
    class="miniWindow "
  >
    <div
      class="title"
      style="height: 40px;"
    >
      <div />
      <img
        alt="close"
        class="iconClose"
        src="close.png"
      />
    </div>
    <div
      class="container"
    >
      <div>
        test
      </div>
      <div
        class="toolbar toolbar"
      >
        <div
          title="切换3D视图"
        >
          <div
            class="icon "
            data-selected="false"
            role="button"
            style="background-image: url(3D.png);"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`miniWindow should render dom correctly 2`] = `
<div>
  <div
    class="miniWindow "
  >
    <div
      class="title"
      style="height: 40px;"
    >
      <div />
    </div>
    <div
      class="container"
    >
      <div>
        test
      </div>
      <div
        class="toolbar toolbar"
      >
        <div
          title="切换3D视图"
        >
          <div
            class="icon "
            data-selected="false"
            role="button"
            style="background-image: url(3D.png);"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`miniWindow should render dom correctly 3`] = `
<div>
  <div
    class="miniWindow "
  >
    <div
      class="title"
      style="height: 40px;"
    >
      <div>
        test
      </div>
    </div>
    <div
      class="container"
    >
      <div>
        test
      </div>
      <div
        class="toolbar toolbar"
      >
        <div
          title="切换3D视图"
        >
          <div
            class="icon "
            data-selected="false"
            role="button"
            style="background-image: url(3D.png);"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
